dbms/src/Functions/zCurveBase.h

```diff
@@ -0,0 +1,342 @@
+                field = Field(val);
+            }
+            else if (type_id == TypeIndex::Int8)
+            {
+                Int8 val;
+                memcpy(&val, &src, sizeof(val));
+                field = Field(val);
+            }
+            else if (type_id == TypeIndex::Int16)
+            {
+                Int16 val;
+                memcpy(&val, &src, sizeof(val));
+                field = Field(val);
+            }
+            else if (type_id == TypeIndex::Int32)
+            {
+                Int32 val;
+                memcpy(&val, &src, sizeof(val));
+                field = Field(val);
+            }
+            else if (type_id == TypeIndex::Int64)
+            {
+                Int64 val;
+                memcpy(&val, &src, sizeof(val));
+                field = Field(val);
+            }
+        }
+        /// Extract the i-th argument from given z value,
+        /// which is just taking every arity-th bit starting from some position
+        ResultType extractArgument(ResultType z_value, size_t argument_index, size_t arity) const
+        {
+            static constexpr size_t number_of_bits = (sizeof(ResultType) << 3);
+            ResultType result = 0;
+            ResultType res_bit = 1ull << (number_of_bits - 1);
+            for (ResultType bit = 1ull << (number_of_bits - (arity - argument_index)); bit; bit >>= arity)
+            {
+                if (z_value & bit)
+                {
+                    result |= res_bit;
+                }
+                res_bit >>= 1;
+            }
+            return result;
+        }
+        /// Get maximal and minimal possible bit representation of a given argument
+        /// when the values fall in a given range of z values.
+        /* For the i-th argument of a function with arity k in the range [L, R] we do roughly the following:
+         * Consider L and R as blocks of k bits, such that the bit corresponding to out argument
+         * is the last bit of every block. The first block (highest-order bits) can be smaller than k.
+         * We try to find the maximum value for the last bit of each block, going from the highest bits.
+         * If the current blocks are L_b and R_b we are basically choosing between L_b and L_b + 1, since
+         * it's useless to pick a higher value. Out of those two we want the one, that is odd
+         * (and thus ends with a set bit, which would increase the value of our argument),
+         * while bearing in mind that the value shouldn't be greater than R_b
+         * (this means only that we can't choose L_b + 1 when L_b = R_b)
+         * If the value L_x we choose is strictly (!) less than R_b, then we should set the remainder of R to ones,
+         * since we can put any bits in the next blocks, without R being a constraint.
+         * We proceed similarly when calculating the minimum possible value.
+         */
+        std::pair<ResultType, ResultType> getMinMaxPossibleBitValueOfArgument(
```

Comment:
It took me a few days, but the meaning of this function is still a little vague. Could you please explain it